### ISSUE TITLE
COLUMNS seems incorrect for powerline-go

### ISSUE THREAD

**Issue by rocketraman** [fish-shell/fish-shell#4519]
# Meta-data:

```
# fish --version
fish, version 2.3.1

# uname -a
Linux edison 4.13.5-200.fc26.x86_64 #1 SMP Thu Oct 5 16:53:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

# echo $TERM
xterm-256color
```

Tried without third-party customizations and still have the same issue.

# Problem:

I am attempting to use powerline-go. For some reason, my prompt is truncated. In discussion in a powerline-go issue, the claim by the author is that fish is not reporting COLUMNS correctly, which is causing the prompt truncation. See https://github.com/justjanne/powerline-go/issues/46#issuecomment-341641769 and the previous few comments.

Is it possible to confirm somehow whether the problem is definitely in `fish` or in `powerline-go`?

**Comment by justjanne** [comment]
Within the context of subprocesses spawned by fish_prompt, $COLUMNS is apparently entirely unset.

**Comment by faho** [comment]
$COLUMNS isn't exported - `set -qx COLUMNS; or echo "Not exported"`.

However, it isn't exported by bash or zsh either! Check e.g. `command env | grep COLUMNS` in either of those.

@justjanne: Your test program prints `$COLUMNS is set to: ''` for bash, fish and zsh for me.

Are you sure you don't have `export COLUMNS` somewhere in your configuration for bash/zsh? As soon as it's inherited by a shell, further modifications will touch the exported variable, so

```fish
> set -gx COLUMNS $COLUMNS
> bash
> echo $COLUMNS # prints e.g. 152
# Now change width of terminal
> zsh
> echo $COLUMNS # prints the new size, e.g. 77
```

**Comment by justjanne** [comment]
> Are you sure you don't have export COLUMNS somewhere in your configuration for bash/zsh? As soon as it's inherited by a shell, further modifications will touch the exported variable, so

Actually, that’s what I had recommended to users if the syscall couldn’t determine terminal width.

But users of Fish report that they are unable to re-export $COLUMNS there.

**Comment by faho** [comment]
So, I've had a deeper look into this.

@rocketraman: Does this work for the second prompt in a session? After you start fish, execute some command (e.g. `ls` or `echo banana`). Is the second prompt truncated as well?

If the second prompt isn't truncated, this is fixed by 93dad29ec670ffdcbb99aee082b2091b84d57e48, which will be included in the (now in beta) 2.7.0. That would make this a duplicate of #4141.

>But users of Fish report that they are unable to re-export $COLUMNS there.

While there have been a bunch of changes to this since 2.3.1, `set -gx COLUMNS $COLUMNS` should work.

>Actually, that’s what I had recommended to users if the syscall couldn’t determine terminal width.

@justjanne: You _really_ want to get it via the syscall. The reason being that exporting $COLUMNS/$LINES actually breaks some applications. See #3839.

Now, arguably these applications are broken, but the syscall should work to begin with, and that's also how fish (and other shells) get the terminal size.

**Comment by justjanne** [comment]
> Now, arguably these applications are broken, but the syscall should work to begin with, and that's also how fish (and other shells) get the terminal size.

I prefer the syscall, of course, but in some situations the syscall isn’t available, but `$COLUMNS` is. For example, some situations include users using cygwin, or the same prompt within of powershell on Windows (apparently that’s possible).

**Comment by rocketraman** [comment]
> @rocketraman: Does this work for the second prompt in a session? After you start fish, execute some command (e.g. ls or echo banana). Is the second prompt truncated as well?
> 
> If the second prompt isn't truncated, this is fixed by 93dad29, which will be included in the (now in beta) 2.7.0. That would make this a duplicate of #4141.

I found that issue when I was researching this and had confirmed that #4141 is not the problem here. I have the same issue for the second and subsequent prompts.

**Comment by rocketraman** [comment]
> While there have been a bunch of changes to this since 2.3.1, set -gx COLUMNS $COLUMNS should work.

@faho  Nope:

```
set: Tried to set the special variable “COLUMNS” with the wrong scope
```

**Comment by zanchey** [comment]
These variables were de-electrified in 2be1288cacab7ddeee407d22e752b0a3bfa16e63 which was released in 2.6.0 - before that, these operations fail:

```
$ set -g COLUMNS 100
set: Tried to change the read-only variable “COLUMNS”
$ set -gx COLUMNS 100
set: Tried to set the special variable “COLUMNS” with the wrong scope
```

I can't think of a fix other than upgrading to fish 2.6.0 or later.

**Comment by mqudsi** [comment]
I'm closing this as it seems that it does not affect the current fish release. 

@rocketraman if you still have a problem getting this working with fish 2.7.1 or fish 3.0 beta, please comment and we'll reopen and investigate further.
